library/lua/utils.lua

```diff
@@ -613,6 +614,60 @@ function processArgs(args, validArgs)
     return result
 end
 
+-- processes commandline options according to optionActions and returns all
+-- argument strings that are not options. Options and non-option strings can
+-- appear in any order, and single-letter options that do not take arguments
+-- can be combined into a single option string (e.g. '-abc' is the same as
+-- '-a -b -c' if options 'a' and 'b' do not take arguments.
+--
+-- optionActions is a vector with elements in the following format:
+-- {shortOptionName, longOptionAlias, hasArg=boolean, handler=fn}
+-- shortOptionName and handler are required. If the option takes an argument,
+-- it will be passed to the handler function.
+-- longOptionAlias is optional.
+-- hasArgument defaults to false.
+--
+-- example usage:
+--
+-- local filename = nil
+-- local open_readonly = false
+-- local nonoptions = processArgsGetopt(args, {
+--   {'r', handler=function() open_readonly = true end},
+--   {'f', 'filename', hasArg=true,
+--    handler=function(optarg) filename = optarg end}
+-- })
+--
+-- when args is {'first', '-f', 'fname', 'second'} or, equivalently,
+-- {'first', '--filename', 'fname', 'second'} (note the double dash in front of
+-- the long option alias), then filename will be fname and nonoptions will
+-- contain {'first', 'second'}.
+function processArgsGetopt(args, optionActions)
+    local sh_opts, long_opts = '', {}
+    local handlers = {}
+    for _,optionAction in ipairs(optionActions) do
+        local sh_opt,long_opt = optionAction[1], optionAction[2]
+        if not sh_opt or type(sh_opt) ~= 'string' or #sh_opt ~= 1 then
+            qerror('optionAction missing option letter at index 1')
```

Comment:
IMO these should be changed to `error` because they're dev-facing errors - I'll adjust locally